Add support for activation and apps route, fix unit tests

cmerchant opened the pull request.

Commit:
Add fetching of crednetials and apps

Commit:
update readme

sanakanchha commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
 	};
 
+	self.getApplicationsStream = (options) => {
+    let exec = co(function *() {
```
why generators? can't we achieve this using async/await?

cmerchant replied:
I think it is still based on node 6 and not being transpiled. It is how the entire library is implemented.

sanakanchha commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
+      return yield req.get({
+        headers : headers,
+        pathname : '/v2/devices',
+				query : options,
```
extra indentation here

cmerchant replied:
I was inconsistent with tabs and spaces. I will fix it.

sanakanchha commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
+        headers : headers,
+        pathname : ['/v2/devices/', deviceId.replace(/\:/g, ''), '/activation'].join(''),
+        query : options,
+				rawStream : true
```
extra tabbing here

cmerchant replied:
Looks like I used spaces except on this line :) I will convert it to tabs.

sanakanchha commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
+        headers : headers,
+        pathname : '/v2/devices',
+				query : options,
+				rawStream : true
```
what is a rawStream? Always set to true?

cmerchant replied:
It's the same as the content library. It is necessary in order to get a stream to the file contents.

Commit:
respond to CR comments

rmomii commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
 	};
 
+	self.getApplicationsStream = (options) => {
+		let exec = co(function *() {
```
For both `getApplicationStream` and `getClientCredentialStream`:
+ I would add some validation here for the `options` and `deviceId` (see `getOrder` function below where it handles non-specified input params)) 
+ We are missing callbacks in both functions (`getOrder` can also be used as an example for this). We have been returning `validation.promiseOrCallback()` on all of our functions (see `lib/validation.js`) to ensure we return the appropriate response when dealing with callbacks and promises.

cmerchant replied:
ok, do we really need a callback or would returning a promise be sufficient?

cmerchant replied:
I guess for consistency with the other modules, we can do the callback.

rmomii commented:
@cmerchant looks really clean! I only had some feedback for the validation/callbacks. Otherwise, it looks great.

brozeph commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
+		return exec;
+	};
+
+	self.getClientCredentialsStream = (deviceId, options) => {
```
Are we returning a file stream here (i.e. binary data) or is this really an API call to get an object that represents the credentials? I would strongly recommend that this should be the latter and I would also argue that we don't want to document this method in the `readme.md` as it's not something a typical 3rd party consumer of our APIs would use. Additionally, we should consider how this method is named (see above comment).

cmerchant replied:
It does return a file stream and then the activation app reads the stream. Utlimately the stream is a JSON object when read. Is it possible to just return directly the JSON object instead of reading the stream? If so, I can implement it that way.

cmerchant replied:
Yep, I can just return the body of the html response which is the json object representing the credentials. This makes the code in the activation app simpler. Thanks for the suggestion!

brozeph commented on lib/provision.js:
```diff
@@ -137,6 +137,40 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
 		return validation.promiseOrCallback(exec, callback);
 	};
 
+	self.getApplicationsStream = (options) => {
```
Are we returning an actual file stream here (i.e. the Docker container itself or some .tar.gz)? If not, we should consider a different name for this method because it infers the response is a stream object... additionally, if it is returning a stream of the HTTP response from an API call, we should encapsulate the response so that it is an object or something easy for the consumer to work with as opposed to a stream where the response data has to be parsed, etc.

cmerchant replied:
It will be a yml file. Perhaps I should leave this out until Garris and Deepak are ready and have finalized the return value.

Commit:
respnod to CR

cmerchant commented:
@rmomii @brozeph I think I covered everything mentioned here. I reverted the readme and removed the getApplicationsStream. Deepak and Garris can work on that when ready.

rmomii commented on lib/provision.js:
```diff
@@ -137,22 +137,13 @@ module.exports = function (provisionOptions, ensureAuthHeaders, self) {
+			callback = options;
+			options = undefined;
+		}
 
```
the only other things is the validation for the `deviceId` but other than that it looks good!

cmerchant replied:
@rmomii the validation is there on line 148, just scroll down a bit. I copied how it was done in the other methods.

rmomii replied:
ahh my bad !! ok, approved!
